x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -414,7 +415,12 @@ public ClusterState execute(ClusterState currentState) {
 
                     if (areClusterStatesCompatibleForRebalance(clusterState, currentState)) {
                         isUpdated = true;
-                        return update(currentState, rebalancedMetadata);
+                        ClusterState updatedState = update(currentState, rebalancedMetadata);
+                        if (TrainedModelAssignmentMetadata.fromState(currentState)
+                            .equals(TrainedModelAssignmentMetadata.fromState(updatedState)) == false) {
+                            systemAuditor.info(Messages.getMessage(Messages.INFERENCE_DEPLOYMENT_REBALANCED, reason));
```

Comment:
It's better to be safe and not do the auditing inside the `ClusterStateUpdateTask::execute` when it could be done in `clusterStateProcessed` using the ML utility thread pool. 

The auditor can potentially do a lot of work if it is the first time a message is audited

https://github.com/elastic/elasticsearch/blob/898d84998b081d47ec6a9c2305f9bf4486d69be3/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/notifications/AbstractAuditor.java#L122